app/src/main/java/org/p2p/wallet/user/interactor/UserTokensInteractor.kt

```diff
@@ -1,18 +1,71 @@
 package org.p2p.wallet.user.interactor
 
-import timber.log.Timber
+import java.math.BigDecimal
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.withContext
+import org.p2p.core.dispatchers.CoroutineDispatchers
 import org.p2p.core.token.Token
+import org.p2p.solanaj.core.PublicKey
 import org.p2p.token.service.api.events.manager.TokenServiceEventPublisher
 import org.p2p.token.service.model.TokenServiceNetwork
+import org.p2p.token.service.model.TokenServicePrice
+import org.p2p.wallet.home.model.TokenComparator
+import org.p2p.wallet.user.repository.UserTokensLocalRepository
+import org.p2p.wallet.user.repository.UserTokensRepository
 
-class UserTokensInteractor(private val tokenServiceInteractor: TokenServiceEventPublisher) {
+class UserTokensInteractor(
+    private val tokenServiceInteractor: TokenServiceEventPublisher,
+    private val userTokensLocalRepository: UserTokensLocalRepository,
+    private val userTokensRepository: UserTokensRepository,
+    private val dispatchers: CoroutineDispatchers
+) {
 
-    fun loadUserRates(userTokens: List<Token.Active>) {
-        Timber.i("Loading user rates for ${userTokens.size}")
+    suspend fun loadUserRates(userTokens: List<Token.Active>) {
         val tokenAddresses = userTokens.map { it.mintAddress }
         tokenServiceInteractor.loadTokensPrice(
             networkChain = TokenServiceNetwork.SOLANA,
             addresses = tokenAddresses
         )
     }
+
+    suspend fun loadUserTokens(publicKey: PublicKey): List<Token.Active> {
+        return userTokensRepository.loadUserTokens(publicKey)
+    }
+
+    suspend fun saveUserTokens(tokens: List<Token.Active>) = withContext(dispatchers.io) {
+        val cachedTokens = userTokensLocalRepository.getUserTokens()
```

Comment:
looks like UserInteractor is not Interactor at all
it should not be resposnible for saving and deciding where to save user tokens

Reply:
and i don't think UserTokens need interactor, there no domain logic associated with them